### ISSUE TITLE
MaximumLikelihoodAmplitudeEstimation returning wrong answer from PhaseOracle

### ISSUE THREAD

**Issue by frankharkins** [Qiskit/qiskit-terra#7366]
### Environment

- **Qiskit Terra version**: `0.18.3`
- **Python version**: `3.9.8`
- **Operating system**: OSX

### What is happening?

For some reason, the `MaximumLikelihoodAmplitudeEstimation` is giving a wrong answer when the problem uses a `PhaseOracle` created from a DIMACS file. Seems to work correctly when using a similar, hand-made`QuantumCircuit` as an oracle.

### How can we reproduce the issue?

```
from qiskit import QuantumCircuit
from qiskit.algorithms import EstimationProblem
from qiskit.circuit.library import PhaseOracle, GroverOperator
from qiskit.algorithms import (AmplitudeEstimation,
                               MaximumLikelihoodAmplitudeEstimation)
from qiskit.providers.aer import QasmSimulator
backend = QasmSimulator()
n = 3

# create state preparation operator
state_prep = QuantumCircuit(n)
state_prep.h(range(n))

# create Grover operator from problem file
oracle = PhaseOracle.from_dimacs_file("examples/3sat.dimacs")
grover_op = GroverOperator(oracle)

problem = EstimationProblem(state_prep,
                            [*range(n)],
                            grover_operator=grover_op)

# Correct result
estimator = AmplitudeEstimation(7, quantum_instance=backend)
result = estimator.estimate(problem)
result.estimation
>>> 0.1295244

# Incorrect result!
estimator = MaximumLikelihoodAmplitudeEstimation(7, quantum_instance=backend)
result = estimator.estimate(problem)
result.estimation
>>> 0.4902973909096974
```
I've attached the DIMACS file (with a `.txt` extension so github will accept it): [3sat.dimacs.txt](https://github.com/Qiskit/qiskit-terra/files/7669713/3sat.dimacs.txt)

Replacing `oracle` with:
```
# create Grover operator by hand
oracle = QuantumCircuit(3)
oracle.h(2)
oracle.x(2)
oracle.ccx(0,1,2)
oracle.x(2)
oracle.h(2)
grover_op = GroverOperator(oracle)
```
behaves correctly.

### What should happen?

The final output should be close to `0.125`.

**Comment by Cryoris** [comment]
This is a strange behavior indeed! But it doesn't seem to be related to the `PhaseOracle` itself. Here is a more minimal example to recreate the same bug
```python
from qiskit import QuantumCircuit
from qiskit.algorithms import EstimationProblem
from qiskit.circuit.library import PhaseOracle, GroverOperator
from qiskit.algorithms import MaximumLikelihoodAmplitudeEstimation, IterativeAmplitudeEstimation
from qiskit.providers.aer import StatevectorSimulator

backend = StatevectorSimulator()
n = 2

# create state preparation operator
state_prep = QuantumCircuit(n)
state_prep.h(range(n))

# create Grover operator from problem file
oracle = QuantumCircuit(n)
oracle.x(1)  # w/o the two X gates around the target qubit, MLAE gives the correct result
oracle.cz(0,1)
oracle.x(1)
grover_op = GroverOperator(oracle)

problem = EstimationProblem(state_prep,
                            [*range(n)],
                            grover_operator=grover_op)

# Correct result
estimator = IterativeAmplitudeEstimation(0.01, 0.05, quantum_instance=backend)
result1 = estimator.estimate(problem)
print(result1.estimation)

# Incorrect result!
estimator = MaximumLikelihoodAmplitudeEstimation(3, quantum_instance=backend)
result2 = estimator.estimate(problem)
print(result2.estimation)
```

**Comment by frankharkins** [comment]
@Cryoris thanks for looking into this! I see now I added some useless X-gates, but there still should be only one solution. If I run your second example I get:
```
0.12499999999999994
0.002667906599823475
```
which is still wrong :/

I took the dimacs file from the [textbook](https://qiskit.org/textbook/ch-applications/satisfiability-grover.html) and added a couple of lines to eliminate the other solutions.
